---
output: github_document
---
<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. `devtools::build_readme()` is handy for this. -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# mmer - The Multimodal Marginal Epistasis test R implementation <img src="man/figures/logo.png" align="right" height="200" alt="" />


<!-- badges: start -->
<!-- badges: end -->

The `mmer` package implements a computationally and statistically efficient method for detecting marginal epistasis in genome-wide association studies (GWAS). 
Find the full package documentation including examples and articles here: [Multimodal Marginal Epistasis test Documentation](https://lcrawlab.github.io/mmer/).

The method builds on

- the Marginal Epistasis Test (MAPIT)
- the multivariate Marginal Epistasis Test (mvMAPIT) 
- the Fast Marginal Epistasis Test (FAME)

The R package documentation for mvMAPIT including examples and articles can be found here: [Multivariate MAPIT Documentation](https://lcrawlab.github.io/mvMAPIT/). 


## Key Features

-	Stochastic Trace Estimation: Enables computational speedup to scale to Biobank scale data.
-	Multimodal Input: Incorporates prior data from HDF5 files to improve power in detecting gene-by-gene interactions.
-	Optimize for Memory Constraints: Highly configurable block wise processing of the data allows to make the most of available resources. See also [How To Optimize the Memory Requirements of MME](articles/tutorial-memory-optimization.html).
-	Parallelization: Utilizes OpenMP for multi-threaded processing.

## Installation

You can install the development version of mmer from [GitHub](https://github.com/) with:

``` r
install.packages("devtools")
devtools::install_github("lcrawlab/mmer")
```

## Dependencies

The full list of R dependencies can be found in the [DESCRIPTION file](https://github.com/lcrawlab/mmer/blob/main/DESCRIPTION).

This package depends on several C++ libraries to provide efficient functionality and performance. 
Make sure the following software are installed and available on your system:

1. HDF5
2. Boost
3. OpenMP

The package ships with the header-only HDF55 C++ library [HighFive](https://github.com/BlueBrain/HighFive).
It utilizes the  library for efficient and index based access to masking data.

## References

- Stamp J, Crawford L (2024). mvMAPIT: Multivariate Genome Wide Marginal Epistasis Test. R package version 2.0.3, https://lcrawlab.github.io/mvMAPIT/, https://github.com/lcrawlab/mvMAPIT.
-	Stamp et al. (2023): Leveraging genetic correlation between traits for epistasis detection in GWAS. G3: Genes, Genomes, Genetics.
- Fu, B., Pazokitoroudi, A., Xue, A., Anand, A., Anand, P., Zaitlen, N., & Sankararaman, S. (2023). A biobank-scale test of marginal epistasis reveals genome-wide signals of polygenic epistasis. bioRxiv.
-	Crawford et al. (2017): Detecting epistasis with the marginal epistasis test. PLoS Genetics.
- Devresse et al. (2024): HighFive - Header-only C++ HDF5 interface. https://zenodo.org/records/13120799

