---
title: "Illustrating the Multimodal Marginal Epistasis Test with Simulated Data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Illustrating the Multimodal Marginal Epistasis Test with Simulated Data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  message = FALSE, 
  warning = FALSE
)
```

The Multimodal Marginal Epistasis test (MME) is designed to condition the genome-wide search for marginal epistasis in quantitative traits on multiple modes of genomic data to both maximize the power of the inference as well as realize the computational efficiency needed to analyze human Biobank scale data. The marginal epistasis framework tests for interactions between a focal variant with any other variant in the data. The multimodal step in the presented method conditions the search for interacting variants with the focal variant on functional genomic data.

This documentation gives a toy example to show case what insights the implementation of the test provides.

```{r setup}
library(mmer)
library(dplyr)
library(ggplot2)
```

```{r run_mme, eval = FALSE}
plink_file <- "path/to/plink/file"
pheno_file <- "path/to/pheno/file"
mask_file <- "path/to/mask/file"
gxg_h5_group <- "gxg"
ld_h5_group <- "ld"
chunksize <- 10
n_randvecs <- 10
n_blocks <- 10
rand_seed <- 123
n_threads <- 5
log_level <- "DEBUG"

n_snps <- 100
snp_indices <- 1:n_snps

mme_result <- mme(
  plink_file,
  pheno_file,
  mask_file,
  snp_indices,
  chunksize,
  n_randvecs,
  n_blocks,
  n_threads,
  gxg_h5_group,
  ld_h5_group,
  rand_seed,
  log_level
)

```

```{r assign_data, include = FALSE}
mme_result <- getting_started
```

```{r}
manhattan_plot <- mme_result$summary %>%
  ggplot(aes(
  x = index,
  y = -log10(p),
  color = true_gxg_snp
)) +
  geom_point() +
  xlab("Position") +
  labs(color = "Epistatic SNP")
plot(manhattan_plot)
```
```{r}
vc_plot <- mme_result$summary %>%
  ggplot(aes(x = true_gxg_snp, y = vc, fill = true_gxg_snp)) +
  geom_boxplot() +
  xlab("True Epistatic SNP") +
  ylab("Phenotypic Variance Explained") +
  labs(color = "Epistatic SNP") +
  theme(legend.position = "none")
plot(vc_plot)
```

